codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/http/HttpBindingGenerator.kt

```diff
@@ -0,0 +1,383 @@
+                    val target = model.expectShape(binding.member.target)
+                    if (target is UnionShape) {
+                        bindEventStreamOutput(operationShape, target)
+                    } else {
+                        deserializeStreamingBody(binding)
+                    }
+                }
+            } else {
+                rustWriter.rustBlock("pub fn $fnName(body: &[u8]) -> std::result::Result<#T, #T>", outputT, errorT) {
+                    deserializePayloadBody(
+                        binding,
+                        errorT,
+                        structuredHandler = structuredHandler,
+                        docShapeHandler = docHandler,
+                        httpMessageType
+                    )
+                }
+            }
+        }
+    }
+
+    private fun RustWriter.bindEventStreamOutput(operationShape: OperationShape, target: UnionShape) {
+        val unmarshallerConstructorFn = EventStreamUnmarshallerGenerator(
+            protocol,
+            model,
+            runtimeConfig,
+            symbolProvider,
+            operationShape,
+            target,
+            mode
+        ).render()
+        rustTemplate(
+            """
+            let unmarshaller = #{unmarshallerConstructorFn}();
+            let body = std::mem::replace(body, #{SdkBody}::taken());
+            Ok(#{Receiver}::new(unmarshaller, body))
+            """,
+            "SdkBody" to RuntimeType.sdkBody(runtimeConfig),
+            "unmarshallerConstructorFn" to unmarshallerConstructorFn,
+            "Receiver" to RuntimeType.eventStreamReceiver(runtimeConfig),
+        )
+    }
+
+    private fun RustWriter.deserializeStreamingBody(binding: HttpBindingDescriptor) {
+        val member = binding.member
+        val targetShape = model.expectShape(member.target)
+        check(targetShape is BlobShape)
+        rustTemplate(
+            """
+            // replace the body with an empty body
+            let body = std::mem::replace(body, #{SdkBody}::taken());
+            Ok(#{ByteStream}::new(body))
+            """,
+            "ByteStream" to RuntimeType.byteStream(runtimeConfig), "SdkBody" to RuntimeType.sdkBody(runtimeConfig)
+        )
+    }
+
+    private fun RustWriter.deserializePayloadBody(
+        binding: HttpBindingDescriptor,
+        errorSymbol: RuntimeType,
+        structuredHandler: RustWriter.(String) -> Unit,
+        docShapeHandler: RustWriter.(String) -> Unit,
+        httpMessageType: HttpMessageType = HttpMessageType.RESPONSE
+    ) {
+        val member = binding.member
+        val targetShape = model.expectShape(member.target)
+        // There is an unfortunate bit of dual behavior caused by an empty body causing the output to be `None` instead
+        // of an empty instance of the response type.
+        withBlock("(!body.is_empty()).then(||{", "}).transpose()") {
+            when (targetShape) {
+                // TODO What if targetShape is set/map/list ?
```

Comment:
This fails the CI: https://github.com/awslabs/smithy-rs/runs/4778596089?check_suite_focus=true#step:4:73

Reply:
Ah, was reading the code and forgot to test that.

Turns out that if you try to target a set/map/list, you get:

```
ERROR: com.amazonaws.simple#HealthcheckInputRequest$payload (ProtocolHttpPayload)
     @ /local/home/davidpz/workplace/smithy-ws/src/SmithyRsSource/codegen-server-test/model/simple.smithy
     |
  36 |     payload: StringList
     |     ^
     = AWS Protocols do not support applying the httpPayload trait to members that target sets, lists, or maps.
```

Which I didn't know, but it's called out in the protocol e.g. [restJson1](https://awslabs.github.io/smithy/1.0/spec/aws/aws-restjson1-protocol.html#supported-traits). We only offer support for AWS protocols:

```
Projection simple failed: software.amazon.smithy.codegen.core.CodegenException: No matching protocol ? service offers: []. We offer: [aws.protocols#awsJson1_0, aws.protocols#awsJson1_1, aws.protocols#awsQuery, aws.protocols#ec2Query, aws.protocols#restJson1, aws.protocols#restXml]
```

So I'll just leave a comment documenting this fact.